samuelfangjw/cadet-frontend
# Interactive SICP-JS Feature

### Description



Add an interactive SICP-JS feature to allow readers to interact with program snippets (with all the playground tools) without taking the reader away from the book. 

#### Key Features & Implementation

##### Navigation
1. Interactive sicp can be accessed via a button in the navbar, which brings users to an index page.
1. Table of contents can be accessed at anytime by clicking the button on the secondary navbar.
1. The previous and next buttons in the secondary navbar navigates to the previous or next page. If the previous or next page does not exist, the buttons should not be rendered.
1. The secondary navbar should not not appear on non-sicp related pages. When an sicp page is viewed, if the academy navbar is present, it should override the academy navbar.

##### Loading a page
1. Preprocessed JSON files are fetched from `source-academy.github.io/sicp/json/XXX.json`.
1. JSON files are parsed into JSX and rendered. 
1. Loading and error components showing different error messages depending on error thrown.

##### Snippets 
1. Snippets in the book that are suitable for evaluation can be clicked. The snippet will expand to reveal the source academy playground. 
    - This implementation uses the actual source academy playground to avoid unnecessary duplication of code. This also means that i have had to add an extra workspace to the redux store.
1. Opening another snippet should close all other snippets. 
     -This was intended as a workaround to having to create multiple workspaces in the redux store, but I ended up preferring it over having each snippet open.
1. Some snippets may have dependencies, in which case one can click on `show/hide dependencies` button to show or hide dependencies respectively. In either case, the program can still be run (dependencies are added to prepend if necessary)

##### Math
Math is parsed using the [better-react-mathjax](https://github.com/fast-reflexes/better-react-mathjax) library. While relatively new and not the most popular library, I could not find a better alternative (though i'm open to ideas!). Requirements: 

     - Support for MathJax 3 or 2.7
     - Typescript support
     - Support for custom delimiters

Other math typesetting libraries such as KaTeX are unable to parse some of the more complex math in SICP, and usually do not support typesetting with math delimiters. An alternative will be to try to use MathJax without a react library, will happily try to change if needed.

##### Refs
Each section is assigned its own ref so that the page can scroll to it. Simply add a hash at the end of the url (e.g. `.../interactive-sicp/1.1.1#p3` scrolls to paragraph 3 once it is rendered.

### Type of change


#### Breaking Changes
1. Minor change to Redux store

```diff
{
    …,
    playground: {
        …,
-       usingSubst: boolean
    },
    workspaces: {
        …
        playground: {
            …
            usingSubst: boolean // previously already defined but unused
        }
    }
}
```

Sagas & reducers have been updated accordingly.

### How to test



#### Manual Testing
1. Set up .env file with location of [sicp web server](https://github.com/source-academy/sicp/wiki/JSON#deployment-to-local-server), or leave blank to use the default github.io one. 
2. Source Academy navbar should have a top level link to the index page of SICP. 
3. Perform acceptance testing of the features above.

#### Also To Test (not yet tested, will tick off once tested)
- [ ] Github and Google persistence features work in SICP
- [ ] Navbar does not have issues when academy/github is active 
- [ ] Check code snippet on other browsers in case they don't display properly
- [ ] Playground works as expected, and state is not affected by sicp snippets
- [ ] No errors thrown on each page

Note: Some minor display bugs may have already been addressed in this [PR](https://github.com/source-academy/sicp/pull/515#issue-662460568) on sicp repo.

#### Automated Testing
Several (although not fully comprehensive) automated tests have also been written, mainly using snapshots to ensure that the component displays correctly.

### Checklist




### Plans for next milestone
- Bookmarks
- Way to generate links to specific sections of text
- Performance (esp in chap 4, some pages with many code and math snippets take too long to load)
- Search feature, either search text or integrate the indexes somehow
